commands/src/main/java/discord4j/commands/CommandBootstrapper.java

```diff
@@ -0,0 +1,163 @@
+package discord4j.commands;
+
+import discord4j.core.DiscordClient;
+import discord4j.core.event.domain.message.MessageCreateEvent;
+import reactor.core.publisher.Flux;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArraySet;
+
+/**
+ * The (thread-safe) entry point for actually using commands. To use, simply instantiate and call
+ * {@link #attach(discord4j.core.DiscordClient)}.
+ *
+ * To register events, obtain the {@link discord4j.commands.CommandDispatcher} via {@link #getCommandDispatcher()}
+ * or create a new {@link discord4j.commands.CommandDispatcher} instance replace the default instance with the new
+ * one via {@link #setCommandDispatcher(CommandDispatcher)}.
+ *
+ * @see #attach(discord4j.core.DiscordClient)
+ * @see discord4j.commands.CommandProvider
+ * @see discord4j.commands.CommandDispatcher
+ * @see discord4j.commands.CommandErrorHandler
+ * @see discord4j.commands.DefaultCommandDispatcher
+ */
+public final class CommandBootstrapper {
+
+    private final static CommandDispatcher DEFAULT_DISPATCHER = new DefaultCommandDispatcher();
+    private final static CommandErrorHandler DEFAULT_ERROR_HANDLER = (context, error) -> {};
+
+    private volatile CommandDispatcher dispatcher = DEFAULT_DISPATCHER;
+    private volatile CommandErrorHandler errorHandler = DEFAULT_ERROR_HANDLER;
+
+    private final Set<CommandProvider> providers;
+
+    /**
+     * Constructs the message listener for commands.
+     *
+     * @see #attach(discord4j.core.DiscordClient)
+     * @param providers The {@link discord4j.commands.CommandProvider}s to initialize this instance with. Note that
+     * this set is expected to be thread-safe as there will be concurrent read access required.
+     */
+    public CommandBootstrapper(Set<CommandProvider> providers) {
+        this.providers = providers;
+    }
+
+    /**
+     * Constructs the message listener for commands.
+     *
+     * @see #attach(discord4j.core.DiscordClient)
+     */
+    public CommandBootstrapper() {
+        this(new CopyOnWriteArraySet<>()); //TODO: LinkedHashSet is probably better, but not concurrent!
```

Comment:
I don't think so. Command loading/unloading is relatively rare, the most expensive this would be is on initialization of all commands typically on startup, but the relative frequency in which commands would load/unload after I feel justifies the acceptable cost copies vs. code mess of having object locks everywhere.

Reply:
Well yeah, that's why I chose CopyOnWriteArraySet. My comment actually refers to the deterministic load order (FIFO) provided by LinkedHashSet which may be desirable in the context of a user manually dealing with this instance

Reply:
It is [ordered](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CopyOnWriteArraySet.html#iterator()).